File: 2023/day21.cpp
// https://adventofcode.com/2023/day/21

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <print>

using garden_t = aoc::char_grid<>;
using positions_t = Vec<point>;

constexpr let plot = '.';
constexpr let rock = '#';

/**
 * This function performs a single step of movement for the pots in the garden.
 *
 * @tparam infinite If true, the garden is considered infinite, meaning that if
 * a pot moves off the edge of the garden, it wraps around to the other side.
 * @param garden    The garden containing the pots.
 * @param starting   The starting positions of the pots.
 * @return           The final positions of the pots after the movement step.
 */
template <bool infinite>
fn move_one(positions_t& ending, garden_t const& garden,
            positions_t const& starting) -> positions_t {
  // Create a vector to store the final positions of the pots.
  ending.clear();
  // Get the dimensions of the garden.
  let bounds = point(garden.num_columns(), garden.num_rows());

  // Loop over each starting pot.
  for (let start : starting) {
    // Loop over each direction a pot can move.
    for (let diff : aoc::basic_neighbor_diffs) {
      // Calculate the position of the neighboring pot.
      let neighbor = start + diff;

      // Check if the garden is not infinite.
      if constexpr (!infinite) {
        // If the garden is not infinite, check if the neighboring pot is in
        // bounds and is a plot. If so, add it to the final positions.
        if (garden.in_bounds(neighbor.y, neighbor.x) &&
            (garden.at(neighbor.y, neighbor.x) == plot)) {
          ending.push_back(neighbor);
        }
      } else {
        // If the garden is infinite, calculate the position of the neighboring
        // pot taking into account wrapping around the edges. Then check if the
        // neighboring pot is a plot. If so, add it to the final positions.
        let neighbor_in_bounds = ((neighbor % bounds) + bounds) % bounds;
        if (garden.at(neighbor_in_bounds.y, neighbor_in_bounds.x) == plot) {
          ending.push_back(neighbor);
        }
      }
    }
  }

  // Sort the final positions in ascending order.
  stdr::sort(ending);
  // Remove any duplicate positions.
  let[non_unique_begin, non_unique_end] = stdr::unique(ending);
  ending.resize(static_cast<usize>(non_unique_begin - ending.begin()));

  // Return the final positions of the pots.
  return std::move(ending);
}

struct input_t {
  garden_t garden;
  positions_t positions;
};

fn parse(String const& filename) -> input_t {
  garden_t garden;
  positions_t positions;

  for (str line : aoc::views::read_lines(filename)) {
    garden.add_row(line);
    let start_pos = line.find('S');
    if (start_pos != str::npos) {
      let start = point(start_pos, garden.num_rows() - 1);
      positions.push_back(start);
      garden.modify(plot, start.y, start.x);
    }
  }

  return input_t{std::move(garden), std::move(positions)};
}

template <int num_steps, bool infinite>
fn solve_case(input_t input) -> u64 {
  auto ending = positions_t{};
  for (i32 i = 0; i < num_steps; ++i) {
    input.positions = move_one<infinite>(ending, input.garden, input.positions);
  }

  return input.positions.size();
}

int main() {
  std::println("Part 1");
  let example = parse("day21.example");
  AOC_EXPECT_RESULT(16, (solve_case<6, false>(example)));
  let input = parse("day21.input");
  AOC_EXPECT_RESULT(3646, (solve_case<64, false>(input)));

  std::println("Part 2");
  AOC_EXPECT_RESULT(16, (solve_case<6, true>(example)));
  AOC_EXPECT_RESULT(50, (solve_case<10, true>(example)));
  AOC_EXPECT_RESULT(1594, (solve_case<50, true>(example)));
  AOC_EXPECT_RESULT(6536, (solve_case<100, true>(example)));
  aoc::return_incomplete();
  AOC_EXPECT_RESULT(167004, (solve_case<500, true>(example)));
  AOC_EXPECT_RESULT(668697, (solve_case<1000, true>(example)));
  AOC_EXPECT_RESULT(16733044, (solve_case<5000, true>(example)));
  AOC_EXPECT_RESULT(1337, (solve_case<26501365, true>(input)));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2023/src/day21.rs
use aoc::grid::{ConfigInput, Grid, Ipos};

type Garden = Grid<char>;
type Pos = Ipos;
type Positions = Vec<Pos>;

const PLOT: char = '.';

struct Input {
    garden: Garden,
    positions: Positions,
}

fn parse(filename: &str) -> Input {
    let (mut garden, config) = Garden::from_file_config(
        filename,
        ConfigInput {
            padding: None,
            start_char: Some('S'),
            end_char: None,
        },
    );

    let start = config.start_pos.unwrap();
    garden.modify(PLOT, start.y, start.x);

    Input {
        garden,
        positions: vec![Pos::new(start.x as isize, start.y as isize)],
    }
}

/// Performs a single step of movement for the pots in the garden.
///
/// If `infinite` is true, the garden wraps around:
/// a pot moving off one edge reappears on the opposite edge.
fn move_one(
    ending: &mut Positions,
    garden: &Garden,
    starting: &Positions,
    infinite: bool,
) -> Positions {
    ending.clear();

    let bounds = Pos::new(garden.num_columns as isize, garden.num_rows as isize);

    for &start in starting {
        for diff in aoc::grid::BASIC_NEIGHBOR_DIFFS {
            let neighbor = start + diff;

            if !infinite {
                if garden.in_bounds_signed(neighbor.y, neighbor.x)
                    && *garden.get(neighbor.y as usize, neighbor.x as usize) == PLOT
                {
                    ending.push(neighbor);
                }
            } else {
                let neighbor_in_bounds = Pos::new(
                    neighbor.x.rem_euclid(bounds.x),
                    neighbor.y.rem_euclid(bounds.y),
                );
                if *garden.get(neighbor_in_bounds.y as usize, neighbor_in_bounds.x as usize) == PLOT
                {
                    ending.push(neighbor);
                }
            }
        }
    }

    ending.sort();
    ending.dedup();

    std::mem::take(ending)
}

fn solve_case(input: &Input, num_steps: i32, infinite: bool) -> u64 {
    let mut positions = input.positions.clone();
    let mut ending = Positions::new();
    for _ in 0..num_steps {
        positions = move_one(&mut ending, &input.garden, &positions, infinite);
    }
    positions.len() as u64
}

fn main() {
    println!("Part 1");
    let example = parse("day21.example");
    aoc::expect_result!(16, solve_case(&example, 6, false));
    let input = parse("day21.input");
    aoc::expect_result!(3646, solve_case(&input, 64, false));

    println!("Part 2");
    aoc::expect_result!(16, solve_case(&example, 6, true));
    aoc::expect_result!(50, solve_case(&example, 10, true));
    aoc::expect_result!(1594, solve_case(&example, 50, true));
    aoc::expect_result!(6536, solve_case(&example, 100, true));
    aoc::return_incomplete();
    aoc::expect_result!(167004, solve_case(&example, 500, true));
    aoc::expect_result!(668697, solve_case(&example, 1000, true));
    aoc::expect_result!(16733044, solve_case(&example, 5000, true));
    aoc::expect_result!(1337, solve_case(&input, 26501365, true));
}
